File: ios/Plugin/ProtocolManager.swift
import Foundation

/**
 * Protocol Manager for HTTP/2 detection and fallback
 */
@objc public class ProtocolManager: NSObject {
    
    // MARK: - Types
    
    struct FallbackConfig {
        var enabled: Bool = true
        var retryCount: Int = 2
        var preferredProtocols: [String] = ["h2", "http/1.1"]
    }
    
    // MARK: - Protocol Detection
    
    /**
     * Detect protocol from URLResponse
     */
    static func detectProtocol(from response: URLResponse) -> String {
        guard let httpResponse = response as? HTTPURLResponse else {
            return "unknown"
        }
        
        // Check for HTTP/2 indicators
        if #available(iOS 13.0, *) {
            // Check for HTTP/2 pseudo-headers
            if httpResponse.value(forHTTPHeaderField: ":status") != nil {
                return "h2"
            }
        }
        
        // Check ALPN negotiated protocol (if available)
        if let url = httpResponse.url,
           let scheme = url.scheme,
           scheme == "https" {
            // Assume HTTP/2 for HTTPS if no clear indicator
            // (URLSession uses HTTP/2 by default for HTTPS)
            return "h2"
        }
        
        return "http/1.1"
    }
    
    /**
     * Check if protocol is supported
     */
    static func isProtocolSupported(_ protocolName: String) -> Bool {
        let normalized = protocolName.lowercased()
        
        switch normalized {
        case "h2", "http/2":
            return true
        case "http/1.1", "http/1.0":
            return true
        case "h3", "http/3":
            // HTTP/3 not natively supported in URLSession yet
            return false
        default:
            return false
        }
    }
    
    /**
     * Get recommended protocols for a given URL
     */
    static func getRecommendedProtocols(for url: String) -> [String] {
        var protocols: [String] = []
        
        if url.starts(with: "https://") {
            // For HTTPS, prefer HTTP/2
            protocols.append("h2")
            protocols.append("http/1.1")
        } else {
            // For HTTP, only use HTTP/1.1
            protocols.append("http/1.1")
        }
        
        return protocols
    }
    
    /**
     * Create URLSessionConfiguration with protocol optimization
     */
    static func createOptimizedConfiguration(
        timeout: TimeInterval = 30.0,
        withCredentials: Bool = false
    ) -> URLSessionConfiguration {
        let config = URLSessionConfiguration.default
        
        // Enable HTTP/2 multiplexing
        config.httpShouldUsePipelining = true
        config.httpMaximumConnectionsPerHost = 6
        
        // Timeout configuration
        config.timeoutIntervalForRequest = timeout
        config.timeoutIntervalForResource = timeout * 2
        
        // Cookie and credential handling
        if withCredentials {
            config.httpCookieAcceptPolicy = .always
            config.httpShouldSetCookies = true
        }
        
        // Network service type for prioritization
        config.networkServiceType = .responsiveData
        
        // TLS configuration for modern protocols
        config.tlsMinimumSupportedProtocolVersion = .TLSv12
        
        // Connection pooling
        config.httpMaximumConnectionsPerHost = 6
        config.requestCachePolicy = .useProtocolCachePolicy
        
        return config
    }
    
    /**
     * Perform request with protocol fallback
     */
    static func performRequestWithFallback(
        request: URLRequest,
        config: FallbackConfig = FallbackConfig(),
        completion: @escaping (Data?, URLResponse?, Error?) -> Void
    ) {
        let session = URLSession(configuration: createOptimizedConfiguration())
        
        performRequestAttempt(
            request: request,
            session: session,
            attempt: 0,
            maxAttempts: config.retryCount,
            completion: completion
        )
    }
    
    private static func performRequestAttempt(
        request: URLRequest,
        session: URLSession,
        attempt: Int,
        maxAttempts: Int,
        completion: @escaping (Data?, URLResponse?, Error?) -> Void
    ) {
        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                // Check if we should retry
                if attempt < maxAttempts {
                    print("[ProtocolManager] Request failed on attempt \(attempt + 1), retrying...")
                    
                    // Exponential backoff
                    let delay = Double(attempt + 1) * 0.1
                    DispatchQueue.global().asyncAfter(deadline: .now() + delay) {
                        performRequestAttempt(
                            request: request,
                            session: session,
                            attempt: attempt + 1,
                            maxAttempts: maxAttempts,
                            completion: completion
                        )
                    }
                } else {
                    // Exhausted retries
                    completion(data, response, error)
                }
            } else {
                // Success
                if let response = response {
                    let protocol = detectProtocol(from: response)
                    print("[ProtocolManager] Request successful with protocol: \(protocol)")
                }
                completion(data, response, error)
            }
        }
        
        task.resume()
    }
    
    /**
     * Get protocol metrics from response
     */
    static func getProtocolMetrics(from response: URLResponse) -> [String: Any] {
        var metrics: [String: Any] = [:]
        
        metrics["protocol"] = detectProtocol(from: response)
        
        if let httpResponse = response as? HTTPURLResponse {
            metrics["statusCode"] = httpResponse.statusCode
            metrics["url"] = httpResponse.url?.absoluteString ?? ""
        }
        
        return metrics
    }
}

File: android/src/main/java/com/capacitor/cors/ProtocolManager.java
package com.capacitor.cors;

import android.util.Log;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import okhttp3.*;

/**
 * Protocol Manager for HTTP/2 and HTTP/3 detection and fallback
 */
public class ProtocolManager {
    private static final String TAG = "ProtocolManager";
    
    /**
     * Protocol fallback configuration
     */
    public static class FallbackConfig {
        public boolean enabled = true;
        public int retryCount = 2;
        public List<Protocol> preferredProtocols = Arrays.asList(
            Protocol.HTTP_2,
            Protocol.HTTP_1_1
        );
        
        public FallbackConfig() {}
        
        public FallbackConfig(boolean enabled, int retryCount, List<Protocol> preferredProtocols) {
            this.enabled = enabled;
            this.retryCount = retryCount;
            this.preferredProtocols = preferredProtocols;
        }
    }
    
    /**
     * Create OkHttpClient with protocol fallback support
     */
    public static OkHttpClient createClientWithFallback(
        OkHttpClient baseClient,
        FallbackConfig config
    ) {
        if (!config.enabled) {
            return baseClient;
        }
        
        return baseClient.newBuilder()
            .addInterceptor(new ProtocolFallbackInterceptor(config))
            .build();
    }
    
    /**
     * Protocol Fallback Interceptor
     */
    private static class ProtocolFallbackInterceptor implements Interceptor {
        private final FallbackConfig config;
        
        public ProtocolFallbackInterceptor(FallbackConfig config) {
            this.config = config;
        }
        
        @Override
        public Response intercept(Chain chain) throws IOException {
            Request request = chain.request();
            Response response = null;
            IOException lastException = null;
            
            // Try each protocol in order
            for (int attempt = 0; attempt <= config.retryCount; attempt++) {
                try {
                    response = chain.proceed(request);
                    
                    // Log successful protocol
                    Protocol protocol = response.protocol();
                    Log.d(TAG, "Request successful with protocol: " + protocol);
                    
                    return response;
                    
                } catch (IOException e) {
                    lastException = e;
                    Log.w(TAG, "Request failed on attempt " + (attempt + 1) + ": " + e.getMessage());
                    
                    // If we've exhausted retries, throw the exception
                    if (attempt >= config.retryCount) {
                        throw e;
                    }
                    
                    // Wait before retry
                    try {
                        Thread.sleep(100 * (attempt + 1)); // Exponential backoff
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        throw e;
                    }
                }
            }
            
            // Should not reach here, but throw last exception if we do
            if (lastException != null) {
                throw lastException;
            }
            
            throw new IOException("Request failed after all retry attempts");
        }
    }
    
    /**
     * Detect protocol from response
     */
    public static String detectProtocol(Response response) {
        Protocol protocol = response.protocol();
        
        switch (protocol) {
            case HTTP_2:
                return "h2";
            case HTTP_1_1:
                return "http/1.1";
            case HTTP_1_0:
                return "http/1.0";
            default:
                return protocol.toString();
        }
    }
    
    /**
     * Check if protocol is supported
     */
    public static boolean isProtocolSupported(String protocolName) {
        try {
            switch (protocolName.toLowerCase()) {
                case "h2":
                case "http/2":
                    return true;
                case "http/1.1":
                case "http/1.0":
                    return true;
                case "h3":
                case "http/3":
                    return false; // HTTP/3 requires OkHttp 5.0+ (alpha)
                default:
                    return false;
            }
        } catch (Exception e) {
            Log.e(TAG, "Error checking protocol support", e);
            return false;
        }
    }
    
    /**
     * Get recommended protocols for a given URL
     */
    public static List<Protocol> getRecommendedProtocols(String url) {
        List<Protocol> protocols = new ArrayList<>();
        
        // For HTTPS, try HTTP/2
        if (url.startsWith("https://")) {
            protocols.add(Protocol.HTTP_2);
            protocols.add(Protocol.HTTP_1_1);
        } else {
            // For HTTP, only use HTTP/1.1
            protocols.add(Protocol.HTTP_1_1);
        }
        
        return protocols;
    }
}